src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -13,28 +16,102 @@ public WhenKeywordRecommender()
         {
         }
 
-        protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
+        protected override bool IsValidContext(int position, CSharpSyntaxContext context,
+            CancellationToken cancellationToken)
         {
-            if (context.IsCatchFilterContext)
+            return context.IsCatchFilterContext ||
+                (IsAfterCompleteExpressionOrPatternInCaseLabel(context, out var expressionOrPattern) &&
+                !IsTypeName(expressionOrPattern, context.SemanticModel, cancellationToken));
```

Comment:
might be worth having a comment explaining the second clause.  Also, we tend fo avoid complex binary expressions as they're a bit hard to read.  Consider changing to:

```c#
if (isCatchFilter)
{
    return true;
}

if (otherCheck) {
    return true;
}
```

Reply:
#wontfix